googlepay/src/main/java/com/airwallex/android/googlepay/GooglePayComponent.kt

```diff
@@ -63,104 +50,61 @@ class GooglePayComponent : ActionComponent {
-                                listener?.onCompleted(AirwallexPaymentStatus.Cancel)
-                            }
-                        } else {
-                            listener?.onCompleted(AirwallexPaymentStatus.Cancel)
-                        }
-                    } ?: listener?.onCompleted(AirwallexPaymentStatus.Cancel)
+        if (requestCode == ThreeDSecurityActivityLaunch.REQUEST_CODE) {
+            listener?.let {
+                val result = ThreeDSecurityActivityLaunch.Result.fromIntent(data)
+                result?.paymentIntentId?.let { intentId ->
+                    it.onCompleted(AirwallexPaymentStatus.Success(intentId))
                 }
-
-                RESULT_CANCELED -> {
-                    listener?.onCompleted(AirwallexPaymentStatus.Cancel)
+                result?.exception?.let { exception ->
+                    it.onCompleted(AirwallexPaymentStatus.Failure(exception))
                 }
-                /**
-                 * At this stage, the user has already seen a popup informing them an error occurred,
-                 * so only logging is required.
-                 *
-                 * @param statusCode will hold the value of any constant from CommonStatusCode or one of the
-                 * WalletConstants.ERROR_CODE_* constants.
-                 * @see [
-                 * Wallet Constants Library](https://developers.google.com/android/reference/com/google/android/gms/wallet/WalletConstants.constant-summary)
-                 */
-                AutoResolveHelper.RESULT_ERROR -> {
-                    AutoResolveHelper.getStatusFromIntent(data)?.let {
-                        AnalyticsLogger.logError(
-                            "googlepay_payment_data_retrieve",
-                            mutableMapOf("code" to it.statusCode.toString()).apply {
-                                putIfNotNull("message", it.statusMessage)
-                            }
-                        )
-                        ConsoleLogger.error(errorTag, String.format("Error code: %d", it.statusCode))
+                return true
+            } ?: return false
+        } else if (requestCode == GooglePayActivityLaunch.REQUEST_CODE) {
+            when (val result = GooglePayActivityLaunch.Result.fromIntent(data)) {
+                GooglePayActivityLaunch.Result.Cancel -> listener?.onCompleted(
+                    AirwallexPaymentStatus.Cancel
+                )
+
+                is GooglePayActivityLaunch.Result.Failure -> listener?.onCompleted(
+                    AirwallexPaymentStatus.Failure(result.exception)
+                )
+
+                is GooglePayActivityLaunch.Result.Success ->
+                    paymentIntentId?.let {
+                        listener?.onCompleted(AirwallexPaymentStatus.Success(it, result.info))
                     }
-                    listener?.onCompleted(AirwallexPaymentStatus.Cancel)
+
+                else -> {
+                    // no op
```

Comment:
Should we return failure here mentioning unexpected result from google pay?

Reply:
This is actually the result from our interim activity, which theoretically shouldn't get here unless there's deserialisation error when unpacking the activity intent.

Reply:
Unexpected result from Google Pay is handled in `GooglePayLauncherActivity`.